crates/autopilot/src/domain/settlement/solution.rs

```diff
@@ -0,0 +1,363 @@
+//! This module defines Solution as originated from a mined transaction
+//! calldata.
+
+use {
+    super::{
+        tokenized,
+        trade::{self, Trade},
+    },
+    crate::{
+        boundary,
+        domain::{
+            auction::{self, order},
+            competition,
+            eth,
+            fee,
+            OrderUid,
+        },
+    },
+    ethcontract::{common::FunctionExt, tokens::Tokenize, U256},
+    std::collections::HashMap,
+};
+
+/// Settlement originated from a calldata of a settlement transaction.
+#[derive(Debug)]
+pub struct Solution {
+    trades: Vec<Trade>,
+    /// Data that was appended to the regular call data of the `settle()` call
+    /// as a form of on-chain meta data. This is used to associate a
+    /// settlement with an auction.
+    auction_id: auction::Id,
+}
+
+impl Solution {
+    /// Number of bytes that may be appended to the calldata to store an auction
+    /// id.
+    const META_DATA_LEN: usize = 8;
+
+    pub fn auction_id(&self) -> auction::Id {
+        self.auction_id
+    }
+
+    pub fn score(
+        &self,
+        prices: &auction::Prices,
+        policies: &HashMap<OrderUid, Vec<fee::Policy>>,
+    ) -> Result<competition::Score, error::Score> {
+        let score = self
+            .trades
+            .iter()
+            .map(|trade| {
+                trade.score(
+                    prices,
+                    policies
+                        .get(trade.order_uid())
+                        .map(|value| value.as_slice())
+                        .unwrap_or_default(),
+                )
+            })
+            .sum::<Result<eth::Ether, trade::Error>>()?;
+        Ok(competition::Score::new(score)?)
+    }
+
+    pub fn native_surplus(&self, prices: &auction::Prices) -> Result<eth::Ether, trade::Error> {
+        self.trades
+            .iter()
+            .map(|trade| trade.native_surplus(prices))
+            .sum()
+    }
+
+    pub fn native_fee(&self, prices: &auction::Prices) -> Result<eth::Ether, trade::Error> {
+        self.trades
+            .iter()
+            .map(|trade| trade.native_fee(prices))
+            .sum()
+    }
+
+    pub fn new(
+        calldata: &eth::Calldata,
+        domain_separator: &eth::DomainSeparator,
+    ) -> Result<Self, Error> {
+        let function = contracts::GPv2Settlement::raw_contract()
```

Comment:
Can this logic and the TradeFlag encoding be moved into `tokenized`? Seems more appropriate there since it's actually looking at the tokenised solution.

Reply:
moved tokenizing part into `tokenized`